File: Eunleelee/2024-01/1월 16일/붕대감기.py
from collections import deque

def solution(bandage, health, attacks):
    attacks = deque(attacks)
    sequence_success = 0
    max_health = health

    for i in range(1, attacks[-1][0] + 1):
      if i == attacks[0][0]:
         health -= attacks[0][1]
         attacks.popleft()
         sequence_success = 0
      
         if health <= 0:
            return -1

      elif health < max_health:
         sequence_success += 1
         health += bandage[1]

         if sequence_success == bandage[0]:
            health += bandage[2]
            sequence_success = 0 
      
      if health > max_health:
         health = max_health
        
    return health


print(solution([5, 1, 5], 30,	[[2, 10], [9, 15], [10, 5], [11, 5]]))

File: yht0827/1월/1월 16일/붕대 감기.java
import java.util.*;

class Solution {
	public int solution(int[] bandage, int health, int[][] attacks) {
		Map<Integer, Integer> atk = new HashMap<>();
		for (int[] attack : attacks)
			atk.put(attack[0], attack[1]);

		int successive = 0;
		int hpmax = health;

		for (int t = 0; t <= attacks[attacks.length - 1][0]; t++) {

			if (atk.containsKey(t)) {
				health -= atk.get(t);
				successive = 0;
			} else {
				health = Math.min(hpmax, health + bandage[1]);
				successive++;

				if (successive == bandage[0]) {
					health = Math.min(hpmax, health + bandage[2]);
					successive = 0;
				}
			}

			if (health <= 0)
				return -1;
		}

		return health;
	}
}